### PR TITLE
Unit Tests for ipam/ippool

### PR DESCRIPTION
**Pull request by vthapar**
Adds UTs for ipam/ippool.go

### CODE REVIEW

**Comment by mangelajo** [comment]
This change is [Reviewable](https://reviewable.io/reviews/submariner-io/submariner/248)

File: pkg/globalnet/controllers/ipam/ippool_test.go [hunk 1]
```diff
@@ -0,0 +1,201 @@
...
+			requestedIp2, _ := pool.RequestIp(service1, requestIp1)
+			Expect(requestedIp2).Should(Equal(requestIp1))
+		})
+	})
+
+	When("All IPs are allocated", func() {
+		pool,_ := NewIpPool("169.254.1.1/30")
+		service1Ip, _ := pool.Allocate(service1)
+		pool.Allocate(pod1)
+		It("IsFull should return true", func() {
```
Comment (tpantelis) [inline pkg/globalnet/controllers/ipam/ippool_test.go]:
```suggestion
                When("IsFull is called", ...
		    It("should return true", func() {
```

File: pkg/globalnet/controllers/ipam/ippool_test.go [hunk 2]
```diff
@@ -0,0 +1,201 @@
...
+			pool.Allocate(service1)
+			Expect(pool.size - len(pool.available)).Should(Equal(1))
+		})
+		It("should not increment allocated IPs pool size if same key is reused", func() {
+			pool.Allocate(service1)
+			Expect(len(pool.allocated)).Should(Equal(1))
+		})
+		It("should mark allocated IP as not available", func() {
+			Expect(pool.IsAvailable(service1Ip)).To(BeFalse())
+		})
```
Comment (tpantelis) [inline pkg/globalnet/controllers/ipam/ippool_test.go]: Add an `It` to verify `IsFull` returns false?

File: pkg/globalnet/controllers/ipam/ippool_test.go [hunk 3]
```diff
@@ -0,0 +1,201 @@
...
+	})
+
+	When("All IPs are allocated", func() {
+		pool,_ := NewIpPool("169.254.1.1/30")
+		service1Ip, _ := pool.Allocate(service1)
+		pool.Allocate(pod1)
+		It("IsFull should return true", func() {
+			Expect(pool.IsFull()).To(BeTrue())
+		})
+		It("requesting any IP should return error", func() {
```
Comment (tpantelis) [inline pkg/globalnet/controllers/ipam/ippool_test.go]:
```suggestion
                When("any IP is requested for an unallocated key", ...
		    It("should return an error", func() {
```

File: pkg/globalnet/controllers/ipam/ippool_test.go [hunk 4]
```diff
@@ -0,0 +1,201 @@
...
+			pod2Ip, err := pool.Allocate(pod2)
+			Expect(err).To(HaveOccurred())
+			Expect(pod2Ip).Should(Equal(""))
+		})
+		It("should return same IP when same key/Ip is requested", func() {
```
Comment (tpantelis) [inline pkg/globalnet/controllers/ipam/ippool_test.go]:
```suggestion
                When("the same IP is requested for an allocated key", ...
		    It("should return the same IP", func() {
```

File: pkg/globalnet/controllers/ipam/ippool_test.go [hunk 5]
```diff
@@ -0,0 +1,201 @@
...
+			requestedIp, _ := pool.RequestIp(service1, service1Ip)
+			Expect(requestedIp).Should(Equal(requestIp1))
+		})
+		It("should return error when different key/Ip is requested", func() {
```
Comment (tpantelis) [inline pkg/globalnet/controllers/ipam/ippool_test.go]:
```suggestion
                When("a different IP is requested for an allocated key", 
		    It("should return an error", func() {
```

File: pkg/globalnet/controllers/ipam/ippool_test.go [hunk 6]
```diff
@@ -0,0 +1,201 @@
...
+		It("IsFull should return true", func() {
+			Expect(pool.IsFull()).To(BeTrue())
+		})
+		It("requesting any IP should return error", func() {
+			pod2Ip, err := pool.Allocate(pod2)
+			Expect(err).To(HaveOccurred())
+			Expect(pod2Ip).Should(Equal(""))
+		})
+		It("should return same IP when same key/Ip is requested", func() {
+			requestedIp, _ := pool.RequestIp(service1, service1Ip)
```
Comment (tpantelis) [inline pkg/globalnet/controllers/ipam/ippool_test.go]: Should check the error here for consistency.

File: pkg/globalnet/controllers/ipam/ippool_test.go [hunk 7]
```diff
@@ -0,0 +1,201 @@
...
+			Expect(len(pool.available) - available).Should(Equal(1))
+		})
+		It("should decrement allocated IP count by 1", func() {
+			Expect(allocated - len(pool.allocated)).Should(Equal(1))
+		})
+		It("should become available for allocation", func() {
+			Expect(pool.IsAvailable(releasedIp)).To(BeTrue())
+		})
+	})
+	When("UnAllocated IP is Released", func() {
```
Comment (tpantelis) [inline pkg/globalnet/controllers/ipam/ippool_test.go]:
```suggestion
	When("Unallocated IP is Released", func() {
```

File: pkg/globalnet/controllers/ipam/ippool_test.go [hunk 8]
```diff
@@ -0,0 +1,201 @@
...
+})
+
+const (
+	requestIp1 = "169.254.1.1"
+	service1 = "default/service1"
+	pod1 = "default/pod1"
+	pod2 = "default/pod2"
+)
+func testPoolCreation() {
+	When("CIDR is 169.254.0.0", func() {
```
[the remark this answers is not in the document]
Comment (vthapar) [reply pkg/globalnet/controllers/ipam/ippool_test.go]: Done

File: pkg/globalnet/controllers/ipam/ippool_test.go [hunk 9]
```diff
@@ -0,0 +1,201 @@
...
+	})
+	When("CIDR Prefix is /24", func() {
+		It("should create IP pool of size 254", func() {
+			pool, err := NewIpPool("169.254.1.0/24")
+			Expect(err).NotTo(HaveOccurred())
+			Expect(len(pool.available)).Should(Equal(254))
+		})
+	})
+	When("CIDR Prefix is /16", func() {
+		It("should create IP pool of size 254", func() {
```
[the remark this answers is not in the document]
Comment (vthapar) [reply pkg/globalnet/controllers/ipam/ippool_test.go]: Done

File: pkg/globalnet/controllers/ipam/ippool_test.go [hunk 10]
```diff
@@ -0,0 +1,201 @@
...
+)
+func testPoolCreation() {
+	When("CIDR is 169.254.0.0", func() {
+		It("should fail to create IP pool with errors", func() {
+			pool, err := NewIpPool("169.254.0.0")
+			Expect(err).To(HaveOccurred())
+			Expect(pool).To(BeNil())
+		})
+	})
+	When("CIDR is 169.254.0.0/33", func() {
```
[the remark this answers is not in the document]
Comment (vthapar) [reply pkg/globalnet/controllers/ipam/ippool_test.go]: Done

File: pkg/globalnet/controllers/ipam/ippool_test.go [hunk 11]
```diff
@@ -0,0 +1,201 @@
...
+			Expect(err).NotTo(HaveOccurred())
+			Expect(len(pool.allocated)).Should(Equal(0))
+		})
+	})
+
+}
+
+func testIpAllocation() {
+	When("Any IP is requested", func() {
+		pool,_ := NewIpPool("169.254.1.1/30")
```
[the remark this answers is not in the document]
Comment (vthapar) [reply pkg/globalnet/controllers/ipam/ippool_test.go]: done

File: pkg/globalnet/controllers/ipam/ippool_test.go [hunk 12]
```diff
@@ -0,0 +1,201 @@
...
+		available := len(pool.available)
+		allocated := len(pool.allocated)
+		service1Ip, err := pool.Allocate(service1)
+
+		It("should not throw any error", func() {
+			Expect(err).NotTo(HaveOccurred())
+		})
+		It("should return an IP", func() {
+			Expect(service1Ip).NotTo(BeNil())
```
[the remark this answers is not in the document]
Comment (vthapar) [reply pkg/globalnet/controllers/ipam/ippool_test.go]: done

**Comment by vthapar** [comment]
Addressed all review comments in next patchset.

File: pkg/globalnet/controllers/ipam/ippool_test.go [hunk 13]
```diff
@@ -0,0 +1,254 @@
...
+			requestedIp, err := pool.RequestIp(pod1, requestIp1)
+
+			It("should return an error", func() {
+				Expect(err).To(HaveOccurred())
+				Expect(requestedIp).Should(Equal(""))
+			})
+		})
+
+	})
+
```
Comment (dfarrell07) [inline pkg/globalnet/controllers/ipam/ippool_test.go]: Nit: Inconsistent newline at end of fn (here there is one)

File: pkg/globalnet/controllers/ipam/ippool_test.go [hunk 14]
```diff
@@ -0,0 +1,254 @@
...
+		It("should not increment available IP count", func() {
+			Expect(len(pool.available)).Should(Equal(available))
+		})
+
+		It("should not decrement allocated IP count", func() {
+			Expect(len(pool.allocated)).Should(Equal(allocated))
+		})
+	})
+}
```
Comment (dfarrell07) [inline pkg/globalnet/controllers/ipam/ippool_test.go]: Nit: Inconsistent newline at end of fn (here there isn't one)

File: pkg/globalnet/controllers/ipam/ippool_test.go [hunk 15]
```diff
@@ -0,0 +1,254 @@
...
+
+func TestIpPool(t *testing.T) {
+	RegisterFailHandler(Fail)
+	RunSpecs(t, "Globalnet/IPAM/IpPool Suite")
+}
```
Comment (dfarrell07) [inline pkg/globalnet/controllers/ipam/ippool_test.go]: Nit: Inconsistent newline at end of fn (here there isn't one)

File: pkg/globalnet/controllers/ipam/ippool_test.go [hunk 16]
```diff
@@ -0,0 +1,254 @@
...
+		})
+	})
+	When("Any pool is created", func() {
+		It("should have allocated pool size of 0", func() {
+			pool, err := NewIpPool("169.254.1.0/16")
+			Expect(err).NotTo(HaveOccurred())
+			Expect(len(pool.allocated)).Should(Equal(0))
+		})
+	})
+
```
Comment (dfarrell07) [inline pkg/globalnet/controllers/ipam/ippool_test.go]: Nit: Inconsistent newline at end of fn (here there is one)

File: pkg/globalnet/controllers/ipam/ippool_test.go [hunk 17]
```diff
@@ -0,0 +1,254 @@
...
+			Expect(err).NotTo(HaveOccurred())
+		})
+
+		It("should return a valid IP for CIDR range", func() {
+			Expect(service1Ip).NotTo(BeNil())
+			_, ipnet, _ := net.ParseCIDR(testCidr)
+			Expect(ipnet.Contains(net.ParseIP(service1Ip))).Should(BeTrue())
+		})
+
+		It("should decrement available IPs pool size by 1", func() {
```
Comment (dfarrell07) [inline pkg/globalnet/controllers/ipam/ippool_test.go]: Nit: Inconsistent naming, this is "IP pool" above and "IPs pool" here (I think the singular makes more sense)

File: pkg/globalnet/controllers/ipam/ippool_test.go [hunk 18]
```diff
@@ -0,0 +1,254 @@
...
+			Expect(available - len(pool.available)).Should(Equal(1))
+		})
+
+		It("should increment allocated IPs pool size by 1", func() {
```
Comment (dfarrell07) [inline pkg/globalnet/controllers/ipam/ippool_test.go]: Nit: Inconsistent naming, this is "IP pool" above and "IPs pool" here (I think the singular makes more sense)

File: pkg/globalnet/controllers/ipam/ippool_test.go [hunk 19]
```diff
@@ -0,0 +1,254 @@
...
+			Expect(len(pool.allocated) - allocated).Should(Equal(1))
+		})
+
+		It("should not decrement available IPs pool size if same key is reused", func() {
```
Comment (dfarrell07) [inline pkg/globalnet/controllers/ipam/ippool_test.go]: Nit: Inconsistent naming, this is "IP pool" above and "IPs pool" here (I think the singular makes more sense)

File: pkg/globalnet/controllers/ipam/ippool_test.go [hunk 20]
```diff
@@ -0,0 +1,254 @@
...
+			pool.Allocate(service1)
+			Expect(pool.size - len(pool.available)).Should(Equal(1))
+		})
+
+		It("should not increment allocated IPs pool size if same key is reused", func() {
```
Comment (dfarrell07) [inline pkg/globalnet/controllers/ipam/ippool_test.go]: Nit: Inconsistent naming, this is "IP pool" above and "IPs pool" here (I think the singular makes more sense)

Comment (vthapar) [reply pkg/globalnet/controllers/ipam/ippool_test.go]: @dfarrell07 Thanks. All addressed.

**Comment by vthapar** [comment]
> I'll take a look at this patch when I find some time. In the meantime, please address the review comments from Tom and the formatting errors (goimports -d main.go pkg test).

Done.

File: pkg/globalnet/controllers/ipam/ippool_test.go [hunk 21]
```diff
@@ -0,0 +1,261 @@
...
+	When("Unallocated IP is Released", func() {
+		pool, _ := NewIpPool(testCidr)
+		_, err := pool.Allocate(service1)
+		available := len(pool.available)
+		allocated := len(pool.allocated)
+		releasedIp := pool.Release(pod1)
+
+		It("should not return an error", func() {
+			Expect(err).NotTo(HaveOccurred())
+		})
```
Comment (sridhargaddam) [inline pkg/globalnet/controllers/ipam/ippool_test.go]: The err we are validating here corresponds to Allocate API. Release() does not return err, so IMO this error validation can be removed. 

Other than this, everything looks good. Nice work!! 
Please squash the commits and we can merge the PR.

Comment (vthapar) [reply pkg/globalnet/controllers/ipam/ippool_test.go]: I get lint errors if I don't store error and then not use it somewhere. That is why had to add it.

Comment (sridhargaddam) [reply pkg/globalnet/controllers/ipam/ippool_test.go]: Okay. Please squash the commits.

**Pull request merged**